### PR TITLE
Infinite loop in wp_nav_menu

### PR DESCRIPTION
**Pull request by peterwilsoncc**
https://core.trac.wordpress.org/ticket/56926

### CODE REVIEW

File: src/wp-includes/nav-menu-template.php [hunk 1]
```diff
@@ -643,3 +629,37 @@ function _nav_menu_item_id_use_once( $id, $item ) {
...
+	}
+
+	/*
+	 * Remove the `menu-item-has-children` class from bottom level menu items.
+	 * -1 is used to display all menu items in one level so the class should
+	 * be removed from all menu items.
+	 */
+	if ( -1 === $max_depth || $depth >= $max_depth ) {
+		$classes = array_diff( $classes, array( 'menu-item-has-children' ) );
+	}
```
Comment (peterwilsoncc) [inline src/wp-includes/nav-menu-template.php:667]: Is `array_diff()` fine or would `array_filter()` be more performant?

Comment (azaozz) [reply src/wp-includes/nav-menu-template.php:667]: > more performant?

Good question. I'll (try to) do a quick test to make sure. Thinking the speed diff between the two won't be significant in this case, probably few nanoseconds. The `$classes` array would usually have around 10 elements if a plugin doesn't do something even "crazier" with styling :)

So perhaps `array_diff()` would be good as it is less code.

Comment (peterwilsoncc) [reply src/wp-includes/nav-menu-template.php:667]: > So perhaps `array_diff()` would be good as it is less code.

This occurred to me this-morning too. As `n = 1` in the second array, there's not much that can be optimized.

**Comment by azaozz** [comment]
Testing this with the steps from https://core.trac.wordpress.org/ticket/56926#comment:21, it prevents the infinite loop in all cases I could think of. 

However noticed that when the menu_item's ID is the same as menu_item's parent, the `menu-item-has-children` class is still added and not removed despite that there are no children (the menu item is set as its own parent, so the rest of the code thinks it has children). 

To fix that thinking the code should reset `$menu_item->menu_item_parent` when it is the same as `$menu_item->ID`. This will prevent the particular case of "bad data" coming from the Customizer error. Code would probably be something like:
```php
	// Set up the $menu_item variables.
	_wp_menu_item_classes_by_context( $menu_items );

	$sorted_menu_items        = array();
	$menu_items_with_children = array();

	foreach ( (array) $menu_items as $menu_item ) {
		// Reset invalid `menu_item_parent`, see: https://core.trac.wordpress.org/ticket/56926.
		if ( (int) $menu_item->ID === (int) $menu_item->menu_item_parent ) {
			$menu_item->menu_item_parent = 0;
		}

		$sorted_menu_items[ $menu_item->menu_order ] = $menu_item;

		if ( $menu_item->menu_item_parent ) {
			$menu_items_with_children[ $menu_item->menu_item_parent ] = true;
		}
	}
```

**Comment by azaozz** [comment]
LGTM, +1 to commit.

**Review by azaozz — APPROVED** [review]

**Comment by peterwilsoncc** [comment]
Merged https://core.trac.wordpress.org/changeset/54999 / https://github.com/WordPress/wordpress-develop/commit/1a8535075613b29e750a3b966db4e2004dfad48d

**Pull request closed without merging**